### PR TITLE
Allow matrix requests with some sources and targets outside of distance limit

### PR DESCRIPTION
**Pull request by acwilton**
This pr should close issue #193 
# Changes
+ When checking the distance between sources and targets in loki, a count of how many requested connections are longer in distance than the max_matrix_distance parameter from the config. If over half of the connections are to long then the request is rejected as usual. Otherwise the request executes and just ignores those connections.
+ An "ignore list" is created in loki and passed to thor. When CostMatrix or TimeDistanceMatrix executes, it takes the ignore list and marks the source + target pairs that are known to be too far as "settled" with a max value that will later be interpreted by the thor service as not found and is output accordingly.
+ These changes do allow matrix queries to have clusters of sources and targets in all different parts of the world in the same query so long as the amount of connections between all sources and targets that are too long do not exceed half of the total amount of requested connections.

### CODE REVIEW

**Comment by kevinkreiser** [comment]
half seems arbitrary. why not just send anything through that doesnt have an ignorelist that is all pairs? i guess my worry would be about how to signal to the user the difference between not finding a path between pairs and them being in the ignore list. maybe we need to send back the ignorelist so that they can tell what happened.

**Comment by acwilton** [comment]
If we send anything through as long as there are some amount of pairs that we could hypothetically find, then I am worried that there would be queries that could be made that could take a very long time. Maybe not though. And my thinking about not marking the points that were ignored any differently was that any pair that comes back with a null value means that it was too far away in cost to calculate so it seems irrelevant whether the point was too far away because our matrix algorithm spent too long looking for it or whether it was because it was past a threshold that we knew from the start would take the matrix algorithm too long to find. It seems more like an internal detail rather than a meaningful one that the user would use. But maybe not. I guess it could be helpful.

**Comment by kevinkreiser** [comment]
summarizing our offline conversation: 

sorry yeah i was under the impression that the path finding part had a limit that was basically dont search forever, give up after some distance, expecting that the distance was greater than the one in loki (as we discussed). seems we may want to fix that but doing so would indeed allow for larger expansion. which brings about the idea of reducing the default limits for things like ped and bike.

on the other hand, for the sake of simplicity maybe it would be fine to just return `null` as normal in either case. i just thought that if we were letting stuff onto thor we would attempt to make ti so that we would get results for those (reducing the likelihood that the limit for expansion in thor is ever hit). to me it doesnt make sense to let stuff through loki only to search to the limit in thor and then send back a null anyway when its just outside the limit. looks like we should have done a bit more diligence here in the past!

File: src/loki/matrix_action.cc [hunk 1]
```diff
@@ -52,8 +66,17 @@ namespace {
...
-          throw valhalla_exception_t{400, 154};
+        if (path_distance > matrix_max_distance) {
+          rapidjson::Value pair(rapidjson::kObjectType);
+          pair.AddMember("source", s, request.GetAllocator());
+          pair.AddMember("target", t, request.GetAllocator());
+
+          ignore_list->PushBack(pair,request.GetAllocator());
+          invalid_count++;
+
+          if ((static_cast<float>(invalid_count) / total_count) > kPercentInvalidAllowed)
```
Comment (dnesbitt61) [inline src/loki/matrix_action.cc]: rather than doing a division at each stage, I would compute a max invalid count outside the loop and compare to that

**Pull request closed without merging**